### PR TITLE
SorrTask596_Extract_email_from_LinkedIn

### PR DESCRIPTION
**Pull request by yiyunlei**
PR for #596. 

Dataset: 
- https://docs.google.com/spreadsheets/d/1TwHK7HgW3-mWvFFiMLYP8u7ANZOjGuae1ajbMID5jo4/edit?usp=sharing

### CODE REVIEW

**Review by gpsaggese — COMMENTED** [review]
Looks good.

Ok to merge as checkpoint, but we need to lint, make a library from it, document the usage. @samarth9008 can you help with this?

We'll test in the Tue's PP.

**Comment by yiyunlei** [comment]
> Looks good.
> 
> Ok to merge as checkpoint, but we need to lint, make a library from it, document the usage. @samarth9008 can you help with this?
> 
> We'll test in the Tue's PP.

I've run the linter and can document the usage later.

File: linkedin/extract_email/email_extraction.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+# Install selenium and openpyxl in the virtual environment.
+install_code = subprocess.call(
+    'sudo /bin/bash -c "(source /venv/bin/activate; pip install selenium)"',
+    shell=True,
+)
+install_code = subprocess.call(
+    'sudo /bin/bash -c "(source /venv/bin/activate; pip install openpyxl)"',
+    shell=True,
+)
+
```
Comment (samarth9008) [inline linkedin/extract_email/email_extraction.py:24]: Can we try doing something like?

```
import helpers.hsystem as hsystem

cmd = 'sudo /bin/bash -c "(source /venv/bin/activate; pip install selenium)"'
hsystem.system(cmd, suppress_output=False, log_level="echo")
```

File: linkedin/extract_email/email_extraction.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    sheet_df = pd.read_excel(filepath, sheet_name=sheet_name)
+    url_list = list(sheet_df["profileUrl"])
+    email_list = get_email_list(url_list, username, password)
+    sheet_df["email"] = email_list
+    with pd.ExcelWriter(filepath, mode="a", if_sheet_exists="replace") as writer:
+        sheet_df.to_excel(writer, sheet_name=sheet_name, index=False)
+
+
+if __name__ == "__main__":
+    # LinkedIn username and password here
```
Comment (samarth9008) [inline linkedin/extract_email/email_extraction.py:77]: -> Add your LinkedIn username and password here.

File: linkedin/extract_email/email_extraction.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    li_username = "your_username"
+    li_password = "your_password"
+    # This is the excel to be read.
```
Comment (samarth9008) [inline linkedin/extract_email/email_extraction.py:80]: -> # Source data file.

File: linkedin/extract_email/email_extraction.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    email_list = get_email_list(url_list, username, password)
+    sheet_df["email"] = email_list
+    with pd.ExcelWriter(filepath, mode="a", if_sheet_exists="replace") as writer:
+        sheet_df.to_excel(writer, sheet_name=sheet_name, index=False)
+
+
+if __name__ == "__main__":
+    # LinkedIn username and password here
+    li_username = "your_username"
+    li_password = "your_password"
```
Comment (samarth9008) [inline linkedin/extract_email/email_extraction.py:79]: -> linkedIn_username
-> linkedIn_password

File: linkedin/extract_email/email_extraction.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    # This is the excel to be read.
+    li_filepath = "Search3-VCs_1st_degree_from_GP.LIn.xlsx"
+    li_sheet_name = "Sheet1"
```
Comment (samarth9008) [inline linkedin/extract_email/email_extraction.py:82]: -> data_file_path
-> data_sheet_name

File: linkedin/extract_email/email_extraction.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        url_list,
+        desc="Processing URLs",
+        position=0,
+        bar_format="{desc}{percentage:3.0f}%|{bar}| {n_fmt}/{total_fmt} [{elapsed}<{remaining}]",
+    ):
+        email_list.append(get_email_from_url(driver, url))
+    return email_list
+
+
+def add_email_to_sheet(filepath, sheet_name, username, password):
```
Comment (samarth9008) [inline linkedin/extract_email/email_extraction.py:67]: Add doc strings and type hints to all the functions.

https://github.com/sorrentum/sorrentum/blob/master/docs/Coding_Style_Guide.md#docstrings-style

https://github.com/sorrentum/sorrentum/blob/master/docs/Coding_Style_Guide.md#use-type-hints

File: linkedin/extract_email/email_extraction.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    sheet_df = pd.read_excel(filepath, sheet_name=sheet_name)
+    url_list = list(sheet_df["profileUrl"])
```
Comment (samarth9008) [inline linkedin/extract_email/email_extraction.py:69]: -> profile_url_list

File: linkedin/extract_email/email_extraction.py (hunk 1 above)
Comment (yiyunlei) [reply linkedin/extract_email/email_extraction.py:24]: Sure.

File: linkedin/extract_email/email_extraction.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env python3
```
Comment (samarth9008) [inline linkedin/extract_email/email_extraction.py:1]: On my side it is giving some errors.

- We are downloading the packages after mentioning them in the import so it is raising import error
```
user_1000@fb0436c07bc3:/app$ python3 linkedin/extract_email/email_extraction.py
Traceback (most recent call last):
  File "/app/linkedin/extract_email/email_extraction.py", line 16, in <module>
    from selenium import webdriver
ModuleNotFoundError: No module named 'selenium'
```
- I installed them manually and ran the script but it is still giving error

```
Traceback (most recent call last):
  File "/app/linkedin/extract_email/email_extraction.py", line 122, in <module>
    email_list = get_email_list(
  File "/app/linkedin/extract_email/email_extraction.py", line 75, in get_email_list
    driver = setup_selenium_linkedin(username, password)
  File "/app/linkedin/extract_email/email_extraction.py", line 35, in setup_selenium_linkedin
    driver = webdriver.Chrome(options=options)
  File "/venv/lib/python3.9/site-packages/selenium/webdriver/chrome/webdriver.py", line 45, in __init__
    super().__init__(
  File "/venv/lib/python3.9/site-packages/selenium/webdriver/chromium/webdriver.py", line 53, in __init__
    self.service.start()
  File "/venv/lib/python3.9/site-packages/selenium/webdriver/common/service.py", line 102, in start
    self.assert_process_still_running()
  File "/venv/lib/python3.9/site-packages/selenium/webdriver/common/service.py", line 115, in assert_process_still_running
    raise WebDriverException(f"Service {self._path} unexpectedly exited. Status code was: {return_code}")
selenium.common.exceptions.WebDriverException: Message: Service /home/.cache/selenium/chromedriver/linux64/119.0.6045.105/chromedriver unexpectedly exited. Status code was: 127
```

Does it occur on your side as well?

Comment (samarth9008) [reply linkedin/extract_email/email_extraction.py:1]: Okay i'll check that

What about import error, I believe the installation should be before the imports

Comment (yiyunlei) [reply linkedin/extract_email/email_extraction.py:1]: Okay, I'll fix this."

Comment (yiyunlei) [reply linkedin/extract_email/email_extraction.py:1]: I've updated and tested the code again; it's working well. You can test it again as well.

**Comment by yiyunlei** [comment]
@gpsaggese @samarth9008 
This code can be discarded since I only tested it a few times (3 to 4 data points each time). My 'fake' LinkedIn account has already been detected as a bot account, which suggests that Selenium can be detected by LinkedIn. Additionally, it can only retrieve emails from first-degree connections.

**Comment by gpsaggese** [comment]
Closing

**Pull request closed without merging**